common/kvstore/tablestore/table_store_builder.go

```diff
@@ -0,0 +1,374 @@
+// Dropping a table is irreversible and will delete all data in the table, so be very careful not to call
+// this method with table names omitted by mistake.
+func (t StoreType) Create(logger logging.Logger, path string, tables ...string) (kvstore.TableStore, error) {
+
+	var base kvstore.Store
+	var err error
+
+	switch t {
+	case LevelDB:
+		base, err = leveldb.NewStore(logger, path)
+		if err != nil {
+			return nil, fmt.Errorf("error creating LevelDB store: %w", err)
+		}
+	case MapStore:
+		base = mapstore.NewStore()
+	default:
+		return nil, fmt.Errorf("unknown store type: %d", t)
+	}
+
+	return create(logger, base, tables...)
+}
+
+// Future work: if we ever decide to permit third parties to provide custom store implementations not in this module,
+// we will need to make this method public.
+
+// create creates a new TableStore instance with the given base store and table names.
+func create(logger logging.Logger, base kvstore.Store, tables ...string) (kvstore.TableStore, error) {
+	metadataTable := newTableView(base, "metadata", metadataTableID)
+	namespaceTable := newTableView(base, "namespace", namespaceTableID)
+
+	err := validateSchema(metadataTable)
+	if err != nil {
+		return nil, fmt.Errorf("error validating schema: %w", err)
+	}
+
+	tableIDMap, err := loadNamespaceTable(namespaceTable)
+	if err != nil {
+		return nil, fmt.Errorf("error loading namespace table: %w", err)
+	}
+
+	err = handleIncompleteDeletion(logger, base, metadataTable, namespaceTable)
+	if err != nil {
+		return nil, fmt.Errorf("error handling incomplete deletion: %w", err)
+	}
+
+	err = addAndRemoveTables(base, metadataTable, namespaceTable, tableIDMap, tables)
+	if err != nil {
+		return nil, fmt.Errorf("error adding and removing tables: %w", err)
+	}
+
+	tableMap := make(map[string]kvstore.Table, len(tableIDMap))
+	for tableID, tableName := range tableIDMap {
+		tableMap[tableName] = newTableView(base, tableName, tableID)
+	}
+	return newTableStore(logger, base, tableMap), nil
+}
+
+// validateSchema loads/initiates the schema version in the metadata table.
+func validateSchema(metadataTable kvstore.Table) error {
+
+	schemaKey := []byte(metadataSchemaVersionKey)
+	onDiskSchemaBytes, err := metadataTable.Get(schemaKey)
+
+	if errors.Is(err, kvstore.ErrNotFound) {
+		// This store is new, no on disk schema version exists.
+		onDiskSchemaBytes = make([]byte, 8)
+		binary.BigEndian.PutUint64(onDiskSchemaBytes, currentSchemaVersion)
+
+		err = metadataTable.Put(schemaKey, onDiskSchemaBytes)
+		if err != nil {
+			return fmt.Errorf("error setting schema version in metadata table: %w", err)
+		}
+
+		return nil
+	} else if err == nil {
+		// Verify schema version.
+		onDiskSchema := binary.BigEndian.Uint64(onDiskSchemaBytes)
+		if onDiskSchema != currentSchemaVersion {
+			// In the future if we change schema versions, we may need to write migration code here.
```

Comment:
How do you migrate? The data will need to be re-processed.

Reply:
Most likely, a migration will require us to iterate all data in the DB and to change its on disk format. The schema version is included here as a "just in case" sort of thing. I REALLY hope we never have to do this, so we should do our best to design the schema right the first time.